### PR TITLE
Megatron-FSDP: shard expert and non-expert parameters by different strategies

### PR DESCRIPTION
**Pull request by rapatel**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?
Lets an MoE model pick a different DP-Shard ZeRO strategy for its expert and non-expert parameters instead of applying one strategy to the whole model, which is what makes combinations like ZeRO-3 on experts with ZeRO-2 on dense weights expressible. Also corrects the offset used to place a main weight shard inside its parameter, which silently corrupted MXFP8 scaling blocks under optim and optim_grads.

:warning: For major changes (either in lines of code or in its impact), please make sure to first share a design doc with the team. If you're unsure what's the best way to do so, contact @NVIDIA/mcore-oncall.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests
- [x] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

**Review by wujingyue — COMMENTED** [review]
Hey @rapatel, thanks for the PR! Can you split this? From the title, it appears to contain three parts naturally.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 
 from .utils import is_torch_min_version
 
+VALID_SHARDING_STRATEGIES = ('no_shard', 'optim', 'optim_grads', 'optim_grads_params')
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py]: " seems to be more common

File: megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py [hunk 2]
```diff
@@ -190,6 +202,17 @@ def __post_init__(self):
         import os
 
         """Check the validity of the config."""
+        for field_name in [
+            "data_parallel_sharding_strategy",
+            "data_parallel_sharding_strategy_experts",
+        ]:
+            strategy = getattr(self, field_name)
+            if strategy is not None and strategy not in VALID_SHARDING_STRATEGIES:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py:217]: What does None even mean? Maybe these fields should default to no_shard instead of None?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 3]
```diff
@@ -85,6 +85,7 @@ def fully_shard_model(
     expt_fsdp_group_ag: Optional[torch.distributed.ProcessGroup] = None,
     fsdp_unit_modules: Optional[Sequence[Type[torch.nn.Module]] | Sequence[str]] = None,
     zero_dp_strategy: str | int = 3,
+    zero_dp_strategy_experts: Optional[str | int] = None,
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]:
```suggestion
    zero_dp_strategy_experts: str | int = 3,
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 4]
```diff
@@ -308,26 +316,30 @@ class that schedules the sharding lifecycle of the model parameters and gradient
...
-        zero_dp_strategy = "optim"
-    elif zero_dp_strategy == ShardingStrategy.OPTIM_GRADS:
-        zero_dp_strategy = "optim_grads"
-    elif zero_dp_strategy == ShardingStrategy.OPTIM_GRADS_PARAMS:
-        zero_dp_strategy = "optim_grads_params"
-    elif zero_dp_strategy in ["no_shard", "optim", "optim_grads", "optim_grads_params"]:
-        # Valid string sharding strategy.
-        pass
-    else:
+    def _parse_zero_dp_strategy(strategy):
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: Can you use match-case?

File: tests/unit_tests/distributed/mfsdp_v1/test_mfsdp_nonuniform_sharding.py
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v1/test_mfsdp_nonuniform_sharding.py:1]: Do nonuniform and hybrid mean different things?

**Review by wujingyue — COMMENTED** [review]
Nice! I'll do another round after this.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py (hunk 1 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py]: Done.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py (hunk 2 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/distributed_data_parallel_config.py:217]: None here means "inherit zero_dp_strategy", not "no_shard". Defaulting it to optim_grads_params would silently change behavior for every existing caller that sets only zero_dp_strategy - a model running uniform zero-1 today would move its experts to zero-3, completely changing the current default behavior. It would also mean the common uniform case has to pass the same strategy twice.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 3 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: Rejecting suggestion - see comment about the consequences of changing this to zero-3 by default. This should remain None.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 4 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: Done.

File: tests/unit_tests/distributed/mfsdp_v1/test_mfsdp_nonuniform_sharding.py
Comment (rapatel) [reply tests/unit_tests/distributed/mfsdp_v1/test_mfsdp_nonuniform_sharding.py:1]: Good catch, renamed to test_mfsdp_nonuniform_sharding.py. This PR's file only covers non-uniform DP-Shard sharding. The HFSDP tests live in test_mfsdp_hybrid_sharding.py in the follow-up PR that adds DP-Outer sharding

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 5]
```diff
@@ -308,26 +316,29 @@ class that schedules the sharding lifecycle of the model parameters and gradient
...
-    else:
-        # Invalid sharding strategy.
-        raise ValueError(
-            f"Invalid FSDP / Inner DP Sharding Strategy: {zero_dp_strategy}\n"
-            f"Valid Sharding Strategies: {ShardingStrategy.NO_SHARD}, "
-            f"{ShardingStrategy.OPTIM}, {ShardingStrategy.OPTIM_GRADS}, "
-            f"{ShardingStrategy.OPTIM_GRADS_PARAMS}, "
-            "no_shard, optim, optim_grads, optim_grads_params"
-        )
+    def _parse_zero_dp_strategy(strategy):
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]:
```suggestion
    def _parse_zero_dp_strategy(strategy: ShardingStrategy | str):
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py [hunk 6]
```diff
@@ -308,26 +316,29 @@ class that schedules the sharding lifecycle of the model parameters and gradient
...
+        # ShardingStrategy is an IntEnum, so the enum members also match the bare ints.
+        match strategy:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:335]:
```
if strategy is a str:
  assert it's a valid strategy str
  return strategy

match strategy:
  ...
```

That said, it feels weird to parse enum into str. It's usually the other way around because str is less type safe.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 5 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py]: Done.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 6 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:335]: Good point - done.

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:335]: > That said, it feels weird to parse enum into str. It's usually the other way around because str is less type safe.

Any ideas on this?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py [hunk 7]
```diff
@@ -366,6 +381,25 @@ def _check_module_parameter_types(self):
                 break
         return expert_params
 
+    def _sharding_strategy_for_param(self, param) -> str:
+        """The DP-Shard sharding strategy of the parameter group owning this parameter."""
+        group_id = self.param_and_grad_buffer.param_to_param_group[param]
+        return self.param_and_grad_buffer.parameter_groups[group_id].sharding_strategy
+
+    def _is_param_fully_sharded(self, param) -> bool:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: I'd inline this to the caller. This helper doesn't dedup a lot but adds an extra indirection, making code less readable. 

Bottom line: optimize for readability, not just conciseness.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py [hunk 8]
```diff
@@ -249,6 +256,28 @@ class BucketingPolicy:
...
     fsdp_unit_modules: List[torch.nn.Module] = dataclasses.field(default_factory=list)
     data_parallel_sharding_strategy: str = "no_shard"
+    expert_data_parallel_sharding_strategy: Optional[str] = None
+
+    def strategy_for(self, is_expert_param: bool) -> str:
+        """Resolve the sharding strategy that applies to a class of parameters."""
+        return get_sharding_strategy(self, is_expert_param)
+
+
+def _buffer_distribution_for_strategy(sharding_strategy: str) -> Tuple[bool, bool, bool]:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: This is related to my other comments. It'd be much better if the input is guaranteed to be an enum for type safety. Otherwise, we'd have to rely on error-prone string matching here.

Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Can this return a NamedTuple so we know which field means what?

File: megatron/training/arguments.py [hunk 9]
```diff
@@ -3198,6 +3144,15 @@ def _add_distributed_args(parser):
     group.add_argument('--data-parallel-sharding-strategy', type=str, default='optim_grads_params',
                        choices=['no_shard', 'optim', 'optim_grads', 'optim_grads_params'],
                        help='Sharding strategy of data parallelism.')
+    group.add_argument('--expert-data-parallel-sharding-strategy', type=str, default=None,
+                       choices=['no_shard', 'optim', 'optim_grads', 'optim_grads_params'],
```
Comment (wujingyue) [inline megatron/training/arguments.py:3326]: The formatter didn't complain about the mixed use of ' and "? Could be another case for a repo-wide style guide.

**Review by wujingyue — APPROVED** [review]

File: megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py (hunk 7 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/megatron_fsdp.py]: Done, inlined.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py (hunk 8 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Agreed, but the string is the source of truth here, not the enum, so guaranteeing an enum at this layer would mean changing the public API:

`DistributedDataParallelConfig.data_parallel_sharding_strategy` is declared `str`, `--data-parallel-sharding-strategy` is an argparse `str` and every consumer compares strings.

So typing this input as an enum means retyping the public config fields, I see there's also a TODO above:
```
# TODO(@cspades): Integrate this Enum into MegatronFSDP.
class ShardingStrategy(IntEnum):
```

Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/param_and_grad_buffer.py]: Good point - done.

File: megatron/training/arguments.py (hunk 9 above)
Comment (rapatel) [reply megatron/training/arguments.py:3326]: Looks like black is configured not to touch quotes in this repo (see `pyproject.toml`:

```
[tool.black]
line_length = 100
skip_string_normalization = true
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py (hunk 6 above)
Comment (rapatel) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/fully_shard.py:335]: Well, I think here `str` is canonical - and enum is really the alias. Remember, the source is actually declared as `str` - `data_parallel_sharding_strategy: str`. So `_parse_zero_dp_strategy` isn't converting the enum into a weaker type, it's normalizing into the one representation.

I agree that flipping it so the enum is canonical is the better option, and I think that's the same TODO Cory suggested - `TODO(@cspades): Integrate this Enum into MegatronFSDP`.

**Review by lmcafee-nvidia — APPROVED** [review]

**Review by Phlip79 — APPROVED** [review]

**Review by guihong-nv — APPROVED** [review]

**Review by Edgarruiz8585 — APPROVED** [review]
{Edgar_Manuel_Ruiz_Arias_217596670, title={Edgar Manuel Ruiz Arias mr ed}, journal=
